skimage/restoration/j_invariant.py

```diff
@@ -0,0 +1,309 @@
+
+    Parameters
+    ----------
+    image : ndarray
+        Input data to be denoised (converted using `img_as_float`).
+    denoise_function : function
+        Denoising function to be calibrated.
+    denoise_parameters : dict of list
+        Ranges of parameters for `denoise_function` to be calibrated over.
+    stride : int, optional
+        Stride used in masking procedure that converts `denoise_function`
+        to J-invariance.
+    multichannel : bool, optional
+        Whether the last axis of the image is to be interpreted as multiple
+        channels or another spatial dimension.
+    approximate_loss : bool, optional
+        Whether to approximate the self-supervised loss used to evaluate the
+        denoiser by only computing it on one masked version of the image.
+        If False, the runtime will be a factor of `stride**image.ndim` longer.
+
+    Returns
+    -------
+    best_denoise_function : function
+        The optimal J-invariant version of `denoise_function`.
+
+    Notes
+    -----
+
+    The calibration procedure uses a self-supervised mean-square-error loss
+    to evaluate the performance of J-invariant versions of `denoise_function`.
+    The minimizer of the self-supervised loss is also the minimizer of the
+    ground-truth loss (i.e., the true MSE error) [1]. The returned function
+    can be used on the original noisy image, or other images with similar
+    characteristics.
+
+    Increasing the stride increases the performance of `best_denoise_function`
+     at the expense of increasing its runtime. It has no effect on the runtime
+     of the calibration.
+
+    References
+    ----------
+    .. [1] J. Batson & L. Royer. Blind Denoising by Self-Supervison.
+           arxiv:1901.11365.
+
+    Examples
+    --------
+
+    >>> from skimage import color, data
+    >>> from skimage.restoration import denoise_wavelet
+    >>> import numpy as np
+    >>> img = color.rgb2gray(data.astronaut()[:50, :50])
+    >>> noisy = img + 0.5 * img.std() * np.random.randn(*img.shape)
+    >>> parameters = {'sigma': np.arange(0.1, 0.4, 0.02)}
+    >>> denoising_function = calibrate_denoiser(noisy, denoise_wavelet,
+    ...                                         denoise_parameters=parameters)
+    >>> denoised_img = denoising_function(img)
+
+    """
+    parameters_tested, losses = calibrate_denoiser_search(
+        image, denoise_function,
+        denoise_parameters=denoise_parameters,
+        stride=stride,
+        multichannel=multichannel,
+        approximate_loss=approximate_loss
+    )
+
+    idx = np.argmin(losses)
+    best_parameters = parameters_tested[idx]
+
+    best_denoise_function = functools.partial(
+        invariant_denoise,
+        denoise_function=denoise_function,
+        stride=stride,
+        multichannel=multichannel,
+        denoiser_kwargs=best_parameters,
+    )
+
+    return best_denoise_function
+
+
+def calibrate_denoiser_search(image, denoise_function, *, denoise_parameters,
```

Comment:
I think it would help if we call this function calibrate_denoiser_search_history. Would this be unacceptably too long?